Source/DafnyCompilers/AST.dfy

```diff
@@ -2,349 +2,349 @@ include "CSharpDafnyASTModel.dfy"
-        MapSelect // Separate node in DafnyAST.cs
-      datatype Datatypes =
-        RankLt | RankGt
-      datatype BinaryOp =
-        | Logical(oLogical: Logical)
-        | Eq(oEq: Eq)
-        | Numeric(oNumeric: Numeric)
-        | BV(oBV: BV)
-        | Char(oChar: Char)
-        | Sequences(oSequences: Sequences)
-        | Sets(oSets: Sets)
-        | Multisets(oMultisets: Multisets)
-        | Maps(oMaps: Maps)
-        | Datatypes(oDatatypes: Datatypes)
-      type T(!new,00,==) = BinaryOp
-    }
-
-    type BinaryOp = BinaryOps.T
-
-    module TernaryOps {
-      import Types
+    type T(!new,00,==) = Type
+  }
+
+  type Type = Types.T
+
+  datatype Tokd<T> =
+    Tokd(tok: Boogie.IToken, val: T)
+
+  module BinaryOps {
+    datatype Logical =
+      Iff // And, Or, and Imp are in LazyOp
+    datatype Eq =
+      EqCommon | NeqCommon
+    datatype Numeric =
+      Lt | Le | Ge | Gt | Add | Sub | Mul | Div | Mod
+    datatype BV =
+      LeftShift | RightShift | BitwiseAnd | BitwiseOr | BitwiseXor
+    datatype Char =
+      LtChar | LeChar | GeChar | GtChar
+    datatype Sequences =
+      SeqEq | SeqNeq | Prefix | ProperPrefix | Concat | InSeq | NotInSeq |
+      SeqSelect | SeqTake | SeqDrop // Separate nodes in DafnyAST.cs
+    datatype Sets =
+      SetEq | SetNeq | Subset | Superset | ProperSubset | ProperSuperset |
+      Disjoint | InSet | NotInSet | Union | Intersection | SetDifference
+    datatype Multisets =
+      MultisetEq | MultisetNeq | MultiSubset | MultiSuperset |
+      ProperMultiSubset | ProperMultiSuperset | MultisetDisjoint | InMultiset |
+      NotInMultiset | MultisetUnion | MultisetIntersection | MultisetDifference |
+      MultisetSelect // Separate node in DafnyAST.cs
+    datatype Maps =
+      MapEq | MapNeq | InMap | NotInMap | MapMerge | MapSubtraction |
+      MapSelect // Separate node in DafnyAST.cs
+    datatype Datatypes =
+      RankLt | RankGt
+    datatype BinaryOp =
+      | Logical(oLogical: Logical)
+      | Eq(oEq: Eq)
+      | Numeric(oNumeric: Numeric)
+      | BV(oBV: BV)
+      | Char(oChar: Char)
+      | Sequences(oSequences: Sequences)
+      | Sets(oSets: Sets)
+      | Multisets(oMultisets: Multisets)
+      | Maps(oMaps: Maps)
+      | Datatypes(oDatatypes: Datatypes)
+    type T(!new,00,==) = BinaryOp
+  }
+
+  type BinaryOp = BinaryOps.T
+
+  module TernaryOps {
+    import Types
+
+    datatype Sequences =
+      SeqUpdate | SeqSubseq
+    datatype Multisets =
+      MultisetUpdate
+    datatype Maps =
+      MapUpdate
+
+    datatype TernaryOp =
+      | Sequences(oSequences: Sequences)
+      | Multisets(oMultisets: Multisets)
+      | Maps(oMaps: Maps)
+
+    type T(!new,00,==) = TernaryOp
+  }
+
+  type TernaryOp = TernaryOps.T
+
+  module UnaryOps {
+    datatype UnaryOp =
+      | BVNot
+      | BoolNot
+      | SeqLength
+      | SetCard
+      | MultisetCard
+      | MapCard
+      | MemberSelect(name: string)
+        // Ghost operators
```

Comment:
wrong indent?

Reply:
Very good finding, but that's expected.
1) We plan to have docstring to be comments like this one right after a constructor, so it's correctly aligned.
2) The comments just below `// |` are aligned with the `|` above because the Dafny formatter recognizes that this is a commented case of a constructor.